// react-authmanager/dist/hoc/auth.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const hoc_manager_1 = __importDefault(require("hoc-manager"));
const react_1 = __importDefault(require("react"));
exports.default = instance => hoc_manager_1.default.create(Component => {
    return class WithAuth extends react_1.default.Component {
        constructor() {
            super(...arguments);
            this.state = {};
        }
        componentWillMount() {
            this.unsubscribe = instance.store.subscribe(() => {
                const { auth } = instance.store.getState();
                if (JSON.stringify(this.state) !== JSON.stringify(auth)) {
                    // @ts-ignore
                    this.setState(auth);
                }
            });
        }
        componentWillUnmount() {
            this.unsubscribe && this.unsubscribe();
        }
        render() {
            const { auth } = instance.store.getState();
            return react_1.default.createElement(Component, Object.assign({}, this.props, { auth: auth, login: data => instance.login(data), logout: () => instance.logout() }));
        }
    };
}, {
    acceptParameters: false
});
